pkg/types/chain_reader.go

```diff
@@ -45,37 +46,40 @@ type ChainReader interface {
 	// Note that implementations should ignore extra fields in params that are not expected in the call to allow easier
 	// use across chains and contract versions.
 	// Similarly, when using a struct for returnVal, fields in the return value that are not on-chain will not be set.
-	GetLatestValue(ctx context.Context, contractName, method string, confidenceLevel primitives.ConfidenceLevel, params, returnVal any) error
+	GetLatestValue(ctx context.Context, readIdentifier string, confidenceLevel primitives.ConfidenceLevel, params, returnVal any) error
 
 	// BatchGetLatestValues batches get latest value calls based on request, which is grouped by contract names that each have a slice of BatchRead.
 	// BatchGetLatestValuesRequest params and returnVal follow same rules as GetLatestValue params and returnVal arguments, with difference in how response is returned.
 	// BatchGetLatestValuesResult response is grouped by contract names, which contain read results that maintain the order from the request.
 	// Contract call errors are returned in the Err field of BatchGetLatestValuesResult.
 	BatchGetLatestValues(ctx context.Context, request BatchGetLatestValuesRequest) (BatchGetLatestValuesResult, error)
 
-	// Bind will override current bindings for the same contract, if one has been set and will return an error if the
-	// contract is not known by the ChainReader, or if the Address is invalid
+	// Bind will add provided bindings and will return an error if the contract is not known by the ContractReader, or if
+	// the Address is invalid. Any provided binding that already exists should result in a noop.
 	Bind(ctx context.Context, bindings []BoundContract) error
 
+	// Unbind will remove all bindings provided.
+	Unbind(ctx context.Context, bindings []BoundContract) error
+
 	// QueryKey provides fetching chain agnostic events (Sequence) with general querying capability.
-	QueryKey(ctx context.Context, contractName string, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]Sequence, error)
+	QueryKey(ctx context.Context, contract BoundContract, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]Sequence, error)
 }
 
 // BatchGetLatestValuesRequest string is contract name.
 type BatchGetLatestValuesRequest map[string]ContractBatch
 type ContractBatch []BatchRead
 type BatchRead struct {
-	ReadName  string
-	Params    any
-	ReturnVal any
+	ReadIdentifier string
+	Params         any
+	ReturnVal      any
 }
```

Comment:
I am scared of this being uninutitive, WDYT about changing  `type BatchGetLatestValuesRequest map[string]ContractBatch` to type `BatchGetLatestValuesRequest map[BoundContract]` and keep BatchRead unchanged. This should keep the api querying more uniform and maybe more intuitive?

Reply:
As long as I add a `String()` method, that should work fine. Seems a bit cleaner that way.

Reply:
The map structure of the GRPC interface has a problem with `map<BoundContract, ContractBatch>`. What if we just went with a simple array in the GRPC interface and the map in the Go type?